graalphp-language/src/main/java/org/graalphp/nodes/array/ArrayWriteNode.java

```diff
@@ -0,0 +1,157 @@
+package org.graalphp.nodes.array;
+
+import com.oracle.truffle.api.dsl.NodeChild;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.library.CachedLibrary;
+import org.graalphp.nodes.PhpExprNode;
+import org.graalphp.runtime.PhpRuntime;
+import org.graalphp.runtime.array.ArrayLibrary;
+import org.graalphp.runtime.array.PhpArray;
+import org.graalphp.util.Logger;
+import org.graalphp.util.PhpLogger;
+
+/**
+ * Array write node uses variable array backends and generalizes if needed
+ *
+ * @author abertschi
+ */
+@NodeChild(value = "receiver")
+@NodeChild(value = "index")
+@NodeChild(value = "value")
+public abstract class ArrayWriteNode extends PhpExprNode {
+
+    private static final Logger L = PhpLogger.getLogger(ArrayWriteNode.class.getSimpleName());
+    private static final boolean DO_TRACE = true;
+
+    public static final String LIMIT = ArrayLibrary.SPECIALIZATION_LIMIT;
+    private static final int DEFAULT_CAPACITY_INCREASE = PhpRuntime.INITIAL_ARRAY_CAPACITY;
+
+    // TODO: the smallest unit for integer numbers are long, java cant have long array length.
+    // we currently cast to int. for better error handling add error handling if not fit within int.
+    // library.write(array.getBackend(), (int) index, value);
+
+    private static void log(String msg) {
+        if (!DO_TRACE) return;
+        L.info(msg);
+    }
+
+    public static ArrayWriteNode create() {
+        return ArrayWriteNodeGen.create(null, null, null);
+    }
+
+    public abstract Object executeWrite(PhpArray array, long index, Object value);
+
+    /**
+     * write within bounds of same type
+     */
+    @Specialization(
+            guards = {
+                    "library.acceptsValue(array.getBackend(), value)"
+                    , "isArrayInBounds(array, index)"
+            },
+            limit = LIMIT)
+    protected PhpArray writeInBoundsSameType(
+            PhpArray array,
+            long index,
+            Object value,
+            @CachedLibrary("array.getBackend()") ArrayLibrary library) {
+        log("writeInBoundsSameType");
+
+        library.write(array.getBackend(), (int) index, value);
+        return array;
+    }
+
+    /**
+     * write within existing capacity of different type => needs generalization
+     */
+    @Specialization(
+            guards = {
+                    "!library.acceptsValue(array.getBackend(), value)"
+                    , "isArrayInBounds(array, index)"
+            },
+            limit = LIMIT)
+    protected PhpArray writeInBoundsWrongType(
```

Comment:
An important design choice here could be that library messages should be as close to language semantics and abstract the representation of the array (in php that is just read + write). You can then internally dispatch on the layout/representation. With this node you are not really abstracting the representation of the array, but just the storage which is only a little bit useful as its actually an implementation detail of the language. But it looks like a good start. 

You could do an ArrayLibrary + StorageLibrary and then let the caches merge by creating a @CachedLibrary to a final field.

Reply:
Thanks a lot for your pointers :+1: Will incorporate it :) Regarding this suggestion, I see the benefits of separation of concern of having a language semantic representation of the array and a technical. Are there other benefits apart from separation of concern, i.e. regarding performance?